python/cudf/cudf/core/udf/groupby_function.py

```diff
@@ -0,0 +1,851 @@
+    extra_args = ", ".join([f"extra_arg_{i}" for i in range(len(args))])
+
+    # Generate the initializers for each device function argument
+    initializers = []
+    for i, (colname, col) in enumerate(frame.items()):
+        idx = str(i)
+        initializers.append(
+            group_initializer_template.format(idx=idx, name=colname)
+        )
+
+    return groupby_apply_kernel_template.format(
+        input_columns=input_columns,
+        extra_args=extra_args,
+        group_initializers="\n".join(initializers),
+    )
+
+
+def _get_groupby_apply_kernel(frame, func, args):
+    dataframe_group_type = _get_frame_groupby_type(
+        np.dtype(list(_all_dtypes_from_frame(frame).items())),
+        frame.index.dtype,
+    )
+
+    return_type = _get_udf_return_type(dataframe_group_type, func, args)
+
+    np_field_types = np.dtype(
+        list(_supported_dtypes_from_frame(frame).items())
+    )
+    dataframe_group_type = _get_frame_groupby_type(
+        np_field_types, frame.index.dtype
+    )
+
+    # Dict of 'local' variables into which `_kernel` is defined
+    global_exec_context = {
+        "cuda": cuda,
+        "Group": Group,
+        "dataframe_group_type": dataframe_group_type,
+        "types": types,
+    }
+    kernel_string = _groupby_apply_kernel_string_from_template(frame, args)
+
+    kernel = _get_kernel_groupby_apply(
+        kernel_string, global_exec_context, func, dev_func_ptx
+    )
+
+    return kernel, return_type
+
+
+@_cudf_nvtx_annotate
+def jit_groupby_apply(offsets, grouped_values, function, *args, cache=True):
+    ngroups = len(offsets) - 1
+
+    if cache is True:
+        kernel, return_type = _compile_or_get(
+            grouped_values, function, args, _get_groupby_apply_kernel
+        )
+    else:
+        kernel, return_type = _get_groupby_apply_kernel(
+            grouped_values, function, args
+        )
+        return_type = numpy_support.as_dtype(return_type)
+
+    output = cp.empty(ngroups, dtype=return_type)
+
+    launch_args = [
+        cp.asarray(offsets),
+        output,
+        cp.asarray(grouped_values.index),
+    ]
+
+    for col in _supported_cols_from_frame(grouped_values).values():
+        launch_args.append(cp.asarray(col))
+
+    launch_args += list(args)
+
+    max_group_size = cp.diff(offsets).max()
+
+    if max_group_size >= 1024:
+        if ngroups < 100:
+            blocklim = 1024
+        else:
+            blocklim = 256
+    else:
+        blocklim = ((max_group_size + 32 - 1) / 32) * 32
+
+    if kernel.specialized:
+        specialized = kernel
+    else:
+        specialized = kernel.specialize(*launch_args)
+
+    # Ask the driver to give a good config
+    ctx = get_context()
+    # Dispatcher is specialized, so there's only one definition - get
+    # it so we can get the cufunc from the code library
+    kern_def = next(iter(specialized.overloads.values()))
+    kwargs = dict(
```

Comment:
I would avoid defining `kwargs` just for one usage. Pass the keyword arguments directly to `ctx.get_max_potential_block_size` instead.